engine/currency_state_manager.go

```diff
@@ -0,0 +1,279 @@
+				if err != nil {
+					log.Errorf(log.ExchangeSys, "%s %s %s: %v",
+						CurrencyStateManagementName,
+						exch.GetName(),
+						enabledAssets[y],
+						err)
+					return
+				}
+
+				// Deploys a full spectrum supported list for the currency states
+				update := map[currency.Code]currencystate.Options{}
+				for x := range pairs {
+					update[pairs[x].Base] = currencystate.Options{}
+					update[pairs[x].Quote] = currencystate.Options{}
+				}
+
+				b := exch.GetBase()
+				if b == nil {
+					log.Errorf(log.ExchangeSys, "%s %s %s: %v",
+						CurrencyStateManagementName,
+						exch.GetName(),
+						enabledAssets[y],
+						"cannot update because base is nil")
+					return
+				}
+				err = b.States.UpdateAll(enabledAssets[y], update)
+				if err != nil {
+					log.Errorf(log.ExchangeSys, "%s %s %s: %v",
+						CurrencyStateManagementName,
+						exch.GetName(),
+						enabledAssets[y],
+						err)
+				}
+				return
+			}
+			log.Errorf(log.ExchangeSys, "%s %s %s: %v",
+				CurrencyStateManagementName,
+				exch.GetName(),
+				enabledAssets[y],
+				err)
+		}
+	}
+}
+
+// CurrencyStateGetAllRPC returns a full snapshot of currency states, whether they
+// are able to be withdrawn, deposited or traded on an exchange for RPC.
+func (c *CurrencyStateManager) CurrencyStateGetAllRPC(exchName string) (*gctrpc.CurrencyStateResponse, error) {
+	if !c.IsRunning() {
+		return nil, fmt.Errorf("%s %w", CurrencyStateManagementName, ErrSubSystemNotStarted)
+	}
+	exch, err := c.GetExchangeByName(exchName)
+	if err != nil {
+		return nil, err
+	}
+
+	sh, err := exch.GetCurrencyStateSnapshot()
+	if err != nil {
+		return nil, err
+	}
+
+	var resp = &gctrpc.CurrencyStateResponse{}
+	for x := range sh {
+		resp.CurrencyStates = append(resp.CurrencyStates, &gctrpc.CurrencyState{
+			Currency:        sh[x].Code.String(),
+			Asset:           sh[x].Asset.String(),
+			WithdrawEnabled: sh[x].Withdraw == nil || *sh[x].Withdraw,
+			DepositEnabled:  sh[x].Deposit == nil || *sh[x].Deposit,
+			TradingEnabled:  sh[x].Trade == nil || *sh[x].Trade,
+		})
+	}
+	return resp, nil
+}
+
+// CurrencyStateWithdraw determines via RPC if the currency code is operational for
+// withdrawal from an exchange for RPC
+func (c *CurrencyStateManager) CurrencyStateWithdrawRPC(exchName string, cc currency.Code, a asset.Item) (*gctrpc.GenericResponse, error) {
+	if !c.IsRunning() {
+		return nil, fmt.Errorf("%s %w", CurrencyStateManagementName, ErrSubSystemNotStarted)
+	}
+
+	exch, err := c.GetExchangeByName(exchName)
+	if err != nil {
+		return nil, err
+	}
+
+	err = exch.CanWithdraw(cc, a)
+	if err != nil {
+		return nil, err
+	}
+	return enabled, nil
+}
+
+// CurrencyStateDeposit determines via RPC if the currency code is operational for
+// depositing to an exchange for RPC
+func (c *CurrencyStateManager) CurrencyStateDepositRPC(exchName string, cc currency.Code, a asset.Item) (*gctrpc.GenericResponse, error) {
+	if !c.IsRunning() {
```

Comment:
I think the test coverage for these functions should be within currency_state_manager_tests.go. The rpcserver_test.go coverage of their functions simply need to call their function and not care about the results.